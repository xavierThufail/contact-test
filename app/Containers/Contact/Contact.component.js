import React, { useState } from 'react';
import { useDispatch, useSelector } from "react-redux";
import { StyleSheet, Text, View, ScrollView, RefreshControl, ActivityIndicator, TouchableOpacity, Image } from 'react-native';
import {BoxShadow} from 'react-native-shadow'

import allAction from '../../Store/actions';
import { Header, Icon, Card } from '../../Components';
import styles from './Contact.style';
import config from './Contact.config';

const renderLoading = () => (
  <View style={styles.containerLoading}>
    <ActivityIndicator 
      size='large'
      color='#11999e'
    />
  </View>
);

const renderRefreshControl = ({ refreshing, onRefresh }) => (
  <RefreshControl refreshing={refreshing} onRefresh={onRefresh} />
);

const renderContacts = ({ contacts }) => contacts.map(contact => <Card {...contact} key={contact.id} />);

const renderScrollView = ({ contacts, refreshing, onRefresh, status }) => (
  <ScrollView
    style={styles.container}
    contentContainerStyle={styles.contentContainer}
    refreshControl={renderRefreshControl({ refreshing, onRefresh })}
  >
    {!!status 
      ? <Text style={{ marginVertical: 5, fontSize: 20, color: status === "contact unavailable" ? '#ea3d13' : '#8fbc5b'}}>{status === "contact unavailable" ? 'Failed to delete Contact' : status}</Text>
      : <Text style={{ marginVertical: 5, color: 'transparent' }}></Text>}
    { contacts && renderContacts({ contacts }) }
  </ScrollView>
);

const ContactComponent = (props) => {
  const { status, loading, contacts, refreshing, onRefresh } = props;

  return (
    <View style={styles.container}>
      <Header title="Contact" />
      
      {loading && contacts.length === 0
        ? renderLoading()
        : renderScrollView({ contacts, refreshing, onRefresh, status })}
    </View>
  );
};

ContactComponent.propTypes = config.component.propTypes;
ContactComponent.displayName = config.component.displayName;

export default ContactComponent;
