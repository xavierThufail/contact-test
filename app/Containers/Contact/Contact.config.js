import PropTypes from 'prop-types';

const component = {
  propTypes: {
    statusDelete: PropTypes.string,
    loading: PropTypes.bool.isRequired,
    contacts: PropTypes.arrayOf(PropTypes.shape({
      id: PropTypes.string,
      firstName: PropTypes.string,
      lastName: PropTypes.string,
      age: PropTypes.number,
      photo: PropTypes.string
    })),
    refreshing: PropTypes.bool.isRequired,
    onRefresh: PropTypes.func.isRequired
  },
  displayName: 'ContactComponent'
};

const container = {
  propTypes: {
    navigation: PropTypes.object,
    route: PropTypes.object,
  },
  displayName: 'ContactContainer'
};

export default {
  container,
  component
};
