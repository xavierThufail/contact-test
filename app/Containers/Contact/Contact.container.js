import React from 'react';

import ContactComponent from './Contact.component';
import useContact from './Contact.hook';
import config from './Contact.config';

const ContactContainer = (props) => {
  const hooks = useContact(props);

  return <ContactComponent {...hooks} />
};

ContactContainer.propTypes = config.container.propTypes;
ContactContainer.displayName = config.container.displayName;

export default ContactContainer;
