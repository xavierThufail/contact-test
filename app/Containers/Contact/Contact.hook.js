import React, { useCallback, useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import allAction from '../../Store/actions';

const useContact = (props) => {
  const { status, loading, contacts } = useSelector(state => state.contact);
  const [refreshing, setRefreshing] = useState(false);
  const dispatch = useDispatch();

  const wait = (timeout) => {
    return new Promise(resolve => {
      setTimeout(resolve, timeout);
    });
  };

  const onRefresh = useCallback(() => {
    setRefreshing(true);
    dispatch(allAction.contact.getContact());
    wait(2000).then(() => setRefreshing(false));
  }, [refreshing]);

  useEffect(() => {
    dispatch(allAction.contact.getContact());
  }, []);

  if (status) {
    setTimeout(() => {
      dispatch(allAction.contact.setStatus(''));
    }, 3000);
  };

  return {
    status,
    loading,
    contacts,
    refreshing,
    onRefresh
  };
};

export default useContact;
