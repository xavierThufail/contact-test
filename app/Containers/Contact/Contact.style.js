const styles = {
  container: {
    flex: 1,
    backgroundColor: '#fff'
  },
  contentContainer: {
    paddingHorizontal: 10,
    paddingBottom: 20,
    paddingTop:0,
    zIndex: 1
  },
  containerLoading: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'center'
  },
  containerImage: {
    height: 50,
    width: 50,
    borderRadius:30,
  },
  row: {
    flexDirection: 'row',
    alignItems: 'center'
  },
  textSubject: {
    marginBottom: 10,
    marginLeft: 10,
    fontSize: 18,
    color: "black",
    fontWeight: 'bold'
  }
};

export default styles;
