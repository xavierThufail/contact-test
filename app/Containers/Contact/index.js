export { default as ContactContainer } from './Contact.container';
