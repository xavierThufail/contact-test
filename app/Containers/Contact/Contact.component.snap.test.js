import React from 'react';
import { render } from '@testing-library/react-native';
import { Provider } from 'react-redux';

import Contact from './Contact.component';
import store from '../../Store';

describe('Contact Component snapshot test', () => {
  it('Should render correctly', () => {
    const mockedData = {
      status: '',
      loading: false,
      contacts: [],
      refreshing: false,
      onRefresh: jest.fn()
    };

    const { toJSON } = render(
      <Provider store={store}>
        <Contact data={mockedData} />
      </Provider>
    );

    expect(toJSON()).toMatchSnapshot()
  })
});
