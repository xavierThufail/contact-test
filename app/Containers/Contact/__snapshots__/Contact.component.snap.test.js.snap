// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Contact Component snapshot test Should render correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginTop": 10,
          "paddingHorizontal": 10,
          "zIndex": 100,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#747474",
            "fontFamily": "source-sans",
            "fontSize": 22,
          }
        }
      >
        Contact
      </Text>
      <View>
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#1b83e3",
              "borderRadius": 10,
              "justifyContent": "center",
              "minHeight": 30,
              "minWidth": 50,
              "opacity": 1,
              "padding": 10,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#f0f0f0",
                "fontWeight": "bold",
                "textAlign": "center",
              }
            }
          >
            Add Contact
          </Text>
        </View>
      </View>
    </View>
  </View>
  <RCTScrollView
    contentContainerStyle={
      Object {
        "paddingBottom": 20,
        "paddingHorizontal": 10,
        "paddingTop": 0,
        "zIndex": 1,
      }
    }
    refreshControl={<RefreshControlMock />}
    style={
      Object {
        "backgroundColor": "#fff",
        "flex": 1,
      }
    }
  >
    <RCTRefreshControl />
    <View>
      <Text
        style={
          Object {
            "color": "transparent",
            "marginVertical": 5,
          }
        }
      />
    </View>
  </RCTScrollView>
</View>
`;
